Introduce a standalone mode

EvanHerman opened the pull request:
## Summary

Addresses issue #1281 

## Relevant technical choices

This PR allows the Google Sitekit plugin to render without any of the WordPress core UI elements (admin bar, admin menu and admin footer).

All that is needed is a query argument `googlesitekit-standalone=1`, and the site kit admin pages will render in a standalone mode. This will work on any page whos name contains `googlesitekit`.

![image](https://user-images.githubusercontent.com/5321364/78707025-0b8bb700-78de-11ea-973a-fe05377868d4.png)

I chose to use `googlesitekit-standalone` instead of `googlesitekit_standalone` just to match the syntax of the page prefix. eg: `googlesitekit-` 

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] We have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

EvanHerman commented:
@googlebot I fixed it.

felixarntz commented on includes/Core/Admin/Standalone.php:
```diff
@@ -0,0 +1,113 @@
+ * Class Google\Site_Kit\Core\Admin\Standalone
+ *
+ * @package   Google\Site_Kit
+ * @copyright 2019 Google LLC
```
Could you turn this into 2020? :)

felixarntz commented on includes/Core/Admin/Standalone.php:
```diff
@@ -0,0 +1,113 @@
+/**
+ * Class managing standlone mode.
+ *
+ * @since 1.0.0
```
Should be `@since n.e.x.t`

felixarntz commented on includes/Core/Admin/Standalone.php:
```diff
@@ -0,0 +1,113 @@
+	/**
+	 * Plugin context.
+	 *
+	 * @since NEXT
```
Should be `@since n.e.x.t` (same for all other such cases).

felixarntz commented on includes/Core/Admin/Standalone.php:
```diff
@@ -0,0 +1,113 @@
+use Google\Site_Kit\Core\Assets\Stylesheet;
+
+/**
+ * Class managing standlone mode.
```
Typo :)

felixarntz commented on includes/Core/Admin/Standalone.php:
```diff
@@ -0,0 +1,113 @@
+	 *
+	 * @since NEXT
+	 */
+	public function standalone_mode() {
```
Let's call this method `register` instead - that's a convention established in all classes in the plugin which add WordPress hooks. The doc-block description should be `Registers functionality through WordPress hooks.`

felixarntz commented on includes/Core/Admin/Standalone.php:
```diff
@@ -0,0 +1,113 @@
+
+		$this->context = $context;
+
+		$this->standalone_mode();
```
This should not be invoked here, but from the `Plugin` class instead (adding hooks from within a constructor is not recommended due to side effects for unit testing).

felixarntz commented on includes/Core/Admin/Standalone.php:
```diff
@@ -0,0 +1,113 @@
+
+		return "{$admin_body_classes} googlesitekit-standalone";
+
+	}
```
Let's remove this public method and make it a closure on the hook above - it shouldn't be public.

felixarntz commented on includes/Core/Admin/Standalone.php:
```diff
@@ -0,0 +1,113 @@
+		);
+
+		$styles->register();
+		$styles->enqueue();
```
Rather than introducing a stylesheet here for these few lines of CSS, I'd prefer if we printed the CSS inline in an `admin_head` callback.

_If_ the decision is to use a CSS file, it should instead be handled by the `Assets` class, and only be enqueued from here, within the `admin_enqueue_scripts` hook. But I think it's not worth doing that for these few lines of CSS, let's print it inline.

EvanHerman replied:
Yeah, I think that inlining those styles make the most sense.

felixarntz commented on includes/Core/Admin/Standalone.php:
```diff
@@ -0,0 +1,113 @@
+	}
+
+	/**
+	 * Append the standalone admin body class.
```
Nitpicking, but we use 3rd person verbs for doc-block descriptions ("Appends...").

felixarntz commented on includes/Core/Admin/Standalone.php:
```diff
@@ -0,0 +1,113 @@
+	}
+
+	/**
+	 * Detect if we are in Google Site Kit standalone mode
```
Same here :)

felixarntz commented on includes/Core/Admin/Standalone.php:
```diff
@@ -0,0 +1,113 @@
+		$standalone = filter_input( INPUT_GET, 'googlesitekit-standalone', FILTER_VALIDATE_BOOLEAN );
+
+		return ( 'admin.php' === $pagenow && ( '' !== $page && false !== strpos( $page, 'googlesitekit' ) ) && $standalone );
+
```
Nit-picking, but could you remove all the blank lines at the beginnings and ends of the method bodies?

felixarntz commented on includes/Plugin.php:
```diff
@@ -157,6 +157,7 @@ function() use ( $options, $transients, $assets ) {
 				( new Core\Admin\Dashboard( $this->context, $assets, $modules ) )->register();
 				( new Core\Notifications\Notifications( $this->context, $options ) )->register();
 				( new Core\Util\Debug_Data( $this->context, $options, $user_options, $authentication, $modules ) )->register();
+				new Core\Admin\Standalone( $this->context );
```
See other comment about how to add hooks in Site Kit, let's call the `register` method on the instance (like in the lines above).

felixarntz commented on includes/Core/Admin/Standalone.php:
```diff
@@ -0,0 +1,113 @@
+		$page       = filter_input( INPUT_GET, 'page', FILTER_SANITIZE_STRING );
+		$standalone = filter_input( INPUT_GET, 'googlesitekit-standalone', FILTER_VALIDATE_BOOLEAN );
+
+		return ( 'admin.php' === $pagenow && ( '' !== $page && false !== strpos( $page, 'googlesitekit' ) ) && $standalone );
```
I think the `'' !== $page` check isn't needed here?

EvanHerman commented:
Thanks for the feedback @felixarntz. Those changes make sense to me. I've refactored things a bit and removed a number of things, which I think clean things up a bit. Feel free to give things another review and I can make the changes ASAP and we can get this pushed through.

felixarntz commented on includes/Core/Admin/Standalone.php:
```diff
@@ -22,6 +22,28 @@
+	/**
+	 * Plugin context.
+	 *
+	 * @since NEXT
```
This is now incorrect again, should be `n.e.x.t` :)

felixarntz commented on includes/Core/Admin/Standalone.php:
```diff
@@ -22,6 +22,28 @@
+	/**
+	 * Constructor.
+	 *
+	 * @since NEXT
```
Same here.

EvanHerman commented:
@felixarntz Doh, sorry about that. Should be fixed now.
